### PR TITLE
v2.0.0

### PR DESCRIPTION
**Pull request by efjacobson**
__BREAKING CHANGES__

* Upgrade to support PHP 7.4.
* Uses `"gdbots/ncr": "^2.1"`
* Uses `"triniti/schemas": "^2.0"`

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -3,14 +3,28 @@
   "description": "Php library that provides implementations for triniti:apollo schemas.",
   "type": "library",
   "license": "Apache-2.0",
+  "repositories": [
+    {
+      "type": "vcs",
+      "url": "https://github.com/triniti/schemas"
+    },
```
Comment (gdbrown) [inline composer.json:10]: @efjacobson this shouldn't be needed, the "dev-blah" is handled by composer automatically.

File: src/CastVoteHandler.php [hunk 2]
```diff
@@ -4,36 +4,24 @@
...
-    use \Gdbots\Ncr\PbjxHelperTrait;
-
-    /**
-     * {@inheritdoc}
-     */
     public static function handlesCuries(): array
     {
         return [
-            CastVoteV1Mixin::findOne()->getCurie(),
+            CastVoteV1::schema()->getCurie(),
```
Comment (gdbrown) [inline src/CastVoteHandler.php:20]: @efjacobson just use strings now

File: src/CreatePollHandler.php [hunk 3]
```diff
@@ -3,22 +3,19 @@
 
 namespace Triniti\Apollo;
 
-use Gdbots\Ncr\AbstractCreateNodeHandler;
+use Gdbots\Ncr\CreateNodeHandler;
+use Gdbots\Pbj\MessageResolver;
 use Gdbots\Pbj\SchemaCurie;
-use Triniti\Schemas\Apollo\Mixin\Poll\PollV1Mixin;
```
Comment (gdbrown) [inline src/CreatePollHandler.php:9]: @efjacobson class not needed, delete it

File: src/DeletePollHandler.php [hunk 4]
```diff
@@ -3,22 +3,19 @@
 
 namespace Triniti\Apollo;
 
-use Gdbots\Ncr\AbstractDeleteNodeHandler;
+use Gdbots\Ncr\DeleteNodeHandler;
+use Gdbots\Pbj\MessageResolver;
 use Gdbots\Pbj\SchemaCurie;
-use Triniti\Schemas\Apollo\Mixin\Poll\PollV1Mixin;
```
Comment (gdbrown) [inline src/DeletePollHandler.php:9]: @efjacobson delete this class

File: src/DeletePollHandler.php [hunk 5]
```diff
@@ -3,22 +3,19 @@
...
 
-class DeletePollHandler extends AbstractDeleteNodeHandler
+class DeletePollHandler extends DeleteNodeHandler
 {
     use PbjxHelperTrait;
```
Comment (gdbrown) [inline src/DeletePollHandler.php:13]: @efjacobson delete this class

File: src/ExpirePollHandler.php [hunk 6]
```diff
@@ -3,22 +3,19 @@
 
 namespace Triniti\Apollo;
```
Comment (gdbrown) [inline src/ExpirePollHandler.php:5]: @efjacobson delete this

File: src/GetPollHistoryRequestHandler.php [hunk 7]
```diff
@@ -3,20 +3,17 @@
...
 namespace Triniti\Apollo;
 
-use Gdbots\Ncr\AbstractGetNodeHistoryRequestHandler;
+use Gdbots\Ncr\GetNodeHistoryRequestHandler;
+use Gdbots\Pbj\MessageResolver;
 use Gdbots\Pbj\SchemaCurie;
-use Triniti\Schemas\Apollo\Mixin\Poll\PollV1Mixin;
 
-class GetPollHistoryRequestHandler extends AbstractGetNodeHistoryRequestHandler
+class GetPollHistoryRequestHandler extends GetNodeHistoryRequestHandler
```
Comment (gdbrown) [inline src/GetPollHistoryRequestHandler.php:10]: @efjacobson delete this class, one will go at site level core for all legacy requests

File: src/GetPollRequestHandler.php [hunk 8]
```diff
@@ -3,22 +3,19 @@
...
-class GetPollRequestHandler extends AbstractGetNodeRequestHandler
+class GetPollRequestHandler extends GetNodeRequestHandler
 {
     use PbjxHelperTrait;
 
-    /**
-     * {@inheritdoc}
-     */
     public static function handlesCuries(): array
     {
```
Comment (gdbrown) [inline src/GetPollRequestHandler.php:15]: @efjacobson class not needed, delete this, done one in site core for legacy requests

File: src/MarkPollAsDraftHandler.php [hunk 9]
```diff
@@ -3,22 +3,19 @@
 
 namespace Triniti\Apollo;
 
-use Gdbots\Ncr\AbstractMarkNodeAsDraftHandler;
+use Gdbots\Ncr\MarkNodeAsDraftHandler;
+use Gdbots\Pbj\MessageResolver;
 use Gdbots\Pbj\SchemaCurie;
-use Triniti\Schemas\Apollo\Mixin\Poll\PollV1Mixin;
```
Comment (gdbrown) [inline src/MarkPollAsDraftHandler.php:9]: @efjacobson delete class

File: src/MarkPollAsPendingHandler.php [hunk 10]
```diff
@@ -3,22 +3,19 @@
...
 namespace Triniti\Apollo;
 
-use Gdbots\Ncr\AbstractMarkNodeAsPendingHandler;
+use Gdbots\Ncr\MarkNodeAsPendingHandler;
+use Gdbots\Pbj\MessageResolver;
 use Gdbots\Pbj\SchemaCurie;
-use Triniti\Schemas\Apollo\Mixin\Poll\PollV1Mixin;
 
-class MarkPollAsPendingHandler extends AbstractMarkNodeAsPendingHandler
+class MarkPollAsPendingHandler extends MarkNodeAsPendingHandler
```
Comment (gdbrown) [inline src/MarkPollAsPendingHandler.php:10]: @efjacobson delete class

File: src/NcrPollProjector.php [hunk 11]
```diff
@@ -3,114 +3,19 @@
...
-use Gdbots\Schemas\Ncr\Mixin\NodeExpired\NodeExpired;
-use Gdbots\Schemas\Ncr\Mixin\NodeMarkedAsDraft\NodeMarkedAsDraft;
-use Gdbots\Schemas\Ncr\Mixin\NodeMarkedAsPending\NodeMarkedAsPending;
-use Gdbots\Schemas\Ncr\Mixin\NodePublished\NodePublished;
-use Gdbots\Schemas\Ncr\Mixin\NodeScheduled\NodeScheduled;
-use Gdbots\Schemas\Ncr\Mixin\NodeUnpublished\NodeUnpublished;
-use Gdbots\Schemas\Ncr\Mixin\NodeUpdated\NodeUpdated;
-use Triniti\Schemas\Apollo\Mixin\Poll\PollV1Mixin;
+use Gdbots\Ncr\NcrProjector;
+use Gdbots\Pbj\MessageResolver;
```
Comment (gdbrown) [inline src/NcrPollProjector.php:8]: @efjacobson delete class

File: src/NcrPollStatsProjector.php [hunk 12]
```diff
@@ -7,115 +7,74 @@
...
-    protected $indexOnReplay = false;
-
-    /**
-     * @param Ncr       $ncr
-     * @param NcrSearch $ncrSearch
-     * @param bool      $indexOnReplay
-     */
     public function __construct(Ncr $ncr, NcrSearch $ncrSearch, bool $indexOnReplay = false)
     {
         $this->ncr = $ncr;
```
Comment (gdbrown) [inline src/NcrPollStatsProjector.php:27]: @efjacobson we'll discuss this one, needs moar

File: src/PbjxHelperTrait.php [hunk 13]
```diff
@@ -3,18 +3,12 @@
...
 trait PbjxHelperTrait
 {
-    /**
-     * @param Node $node
-     *
-     * @return bool
-     */
-    protected function isNodeSupported(Node $node): bool
+    protected function isNodeSupported(Message $node): bool
     {
```
Comment (gdbrown) [inline src/PbjxHelperTrait.php:11]: @efjacobson delete trait

File: src/PollStatsAggregate.php [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+namespace Triniti\Apollo;
+
+use Gdbots\Ncr\Aggregate;
+
+class PollStatsAggregate extends Aggregate
+{
+    /**
+     * This is for legacy uses of command/event mixins for common
+     * ncr operations. It will be removed in 3.x.
+     *
```
Comment (gdbrown) [inline src/PollStatsAggregate.php:13]: @efjacobson probably won't been needed, delete for now

File: src/PublishPollHandler.php [hunk 15]
```diff
@@ -3,22 +3,19 @@
...
 
-class PublishPollHandler extends AbstractPublishNodeHandler
+class PublishPollHandler extends PublishNodeHandler
 {
     use PbjxHelperTrait;
 
-    /**
-     * {@inheritdoc}
-     */
     public static function handlesCuries(): array
```
Comment (gdbrown) [inline src/PublishPollHandler.php:14]: @efjacobson delete class

File: src/SearchPollsRequestHandler.php [hunk 16]
```diff
@@ -3,31 +3,27 @@
...
 class SearchPollsRequestHandler extends AbstractSearchNodesRequestHandler
 {
-    /**
-     * {@inheritdoc}
-     */
     public static function handlesCuries(): array
     {
         return [
-            SearchPollsRequestV1Mixin::findOne()->getCurie(),
+            SearchPollsRequestV1::schema()->getCurie(),
```
Comment (gdbrown) [inline src/SearchPollsRequestHandler.php:22]: @efjacobson use string with new and old, e.g. https://github.com/gdbots/iam-php/blob/master/src/SearchUsersRequestHandler.php#L21

File: src/UnpublishPollHandler.php [hunk 17]
```diff
@@ -3,22 +3,19 @@
...
-class UnpublishPollHandler extends AbstractUnpublishNodeHandler
+class UnpublishPollHandler extends UnpublishNodeHandler
 {
     use PbjxHelperTrait;
 
-    /**
-     * {@inheritdoc}
-     */
     public static function handlesCuries(): array
     {
```
Comment (gdbrown) [inline src/UnpublishPollHandler.php:15]: @efjacobson delete it

File: src/UpdatePollHandler.php [hunk 18]
```diff
@@ -3,22 +3,19 @@
...
 
-class UpdatePollHandler extends AbstractUpdateNodeHandler
+class UpdatePollHandler extends UpdateNodeHandler
 {
     use PbjxHelperTrait;
 
-    /**
-     * {@inheritdoc}
-     */
     public static function handlesCuries(): array
```
Comment (gdbrown) [inline src/UpdatePollHandler.php:14]: @efjacobson canceled

**Review by gdbrown — CHANGES_REQUESTED** [review]

File: src/NcrPollProjector.php (hunk 11 above)
Comment (efjacobson) [reply src/NcrPollProjector.php:8]: @gdbrown i need it for the PollAggregateTest https://github.com/triniti/apollo-php/blob/sy5/tests/PollAggregateTest.php#L40, unless you want to not do that or do it a different way

Comment (gdbrown) [reply src/NcrPollProjector.php:8]: @efjacobson use the projector from gdbots/ncr

**Comment by efjacobson** [comment]
@gdbrown think i got everything, can discuss NcrPollStatsProjector whenever

**Comment by efjacobson** [comment]
@gdbrown lmk know what you think of NcrPollStatsProjector now. fwiw although everything has been consolidated into the single method onPollProjected, i think it is less readable. maybe a good middle ground would be to handle all the similar status events there, with dedicated methods for created, deleted, and vote-casted

**Comment by gdbrown** [comment]
@efjacobson i don't think we need to do any conditional logic for the different events causing the projection.  just take the poll node as it is and apply that.  only oddball is if the item is deleted then delete the stats.  the rest is straight copy of current poll to poll stats.

**Comment by efjacobson** [comment]
@gdbrown like so? https://github.com/triniti/apollo-php/blob/f3802a58c19272474738d0a4e9cf4055c531b315/src/NcrPollStatsProjector.php#L38

will `triniti:apollo:mixin:poll.projected` pick up on the vote-casted projection?

**Comment by gdbrown** [comment]
@efjacobson 
- don't worry about the mixin checking... just check the node status... the event has the node.  `$event->getNode()` is the poll and you can safely use that instead of fetching again later in this method.
- vote casted should remain as its own method and subscribe to that event.

lmk if that makes sense.

**Comment by efjacobson** [comment]
@gdbrown this going in the right direction? https://github.com/triniti/apollo-php/pull/6/commits/be57b207e74c7b0cbc721effa97f7bda170a82ca

**Comment by gdbrown** [comment]
@efjacobson now we're cooking with grease.

**Review by gdbrown — APPROVED** [review]

**Pull request merged**